lib/libzfs_core/libzfs_core.c

```diff
@@ -645,6 +646,79 @@ lzc_send_resume(const char *snapname, const char *from, int fd,
 	    resumeoff, NULL));
 }
 
+void
+lzc_set_pipe_max(int infd)
+{
+#ifdef __linux__
+	FILE *procf = fopen("/proc/sys/fs/pipe-max-size", "re");
+
+	if (procf != NULL) {
+		unsigned long max_psize;
+		long cur_psize;
+		if (fscanf(procf, "%lu", &max_psize) > 0) {
+			cur_psize = fcntl(infd, F_GETPIPE_SZ);
+			if (cur_psize > 0 &&
+			    max_psize > (unsigned long) cur_psize)
+				fcntl(infd, F_SETPIPE_SZ,
+				    max_psize);
+		}
+		fclose(procf);
+	}
+#endif
+}
+
+
+struct sendargs {
+	int ioctlfd;
+	int inputfd;
+	int outputfd;
+};
+typedef struct sendargs sendargs_t;
+
+void *
+__do_send_output(void *voidargs)
+{
+	sendargs_t *args = (sendargs_t *)voidargs;
+	sigset_t sigs;
+	int buflen = 131072;
+
+	// see the comment above the close() calls for why
+	// we can't just die from SIGPIPE
+	sigemptyset(&sigs);
+	sigaddset(&sigs, SIGPIPE);
+	pthread_sigmask(SIG_BLOCK, &sigs, NULL);
+
+#ifndef __linux__
+	void* buf = calloc(1, buflen);
+#endif
+
+	int err = 1;
+	while (err > 0) {
+#ifdef __linux__
+		err = splice(args->inputfd, NULL, args->outputfd, NULL, buflen,
+		    SPLICE_F_MORE);
+#else
+		err = read(args->inputfd, buf, buflen);
+		if (err <= 0) {
+			break;
+		}
+		err = write(args->outputfd, buf, err);
+#endif
+	}
+#ifndef __linux__
+	free(buf);
+#endif
+	if (err < 0) {
+		err = errno;
+	}
+	// If we just return here, the other thread often blocks
+	// indefinitely on the ioctl completing, which won't happen
+	// because we stopped consuming the data. So we close the pipe
+	// here, and the other thread exits in a timely fashion.
```

Comment:
nit: cstyle consistency.

```c
	/*
	 * If we just return here, the other thread often blocks
	 * indefinitely on the ioctl completing, which won't happen
	 * because we stopped consuming the data. So we close the pipe
	 * here, and the other thread exits in a timely fashion.
 	 */
```

Reply:
Done!